### PR TITLE
Packer Builder 1&1

### PR DESCRIPTION
**Pull request by jasminSPC**
This pull request contains code for [1&1](https://www.1and1.com) builder.

### CODE REVIEW

File: website/source/docs/builders/oneandone.html.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+### Optional
+
+-   `disk_size` (string) - Amount of disk space for this image in GB. Defaults to "50"
+
+-   `image_password` (string) - Password for the server images.
+
+-   `image_name` (string) - Resulting image. If "image_name" is not provided Packer will generate it
+
+-   `timeout` (string) - An approximate limit on how long Packer will continue making status requests while waiting for the build to complete. Default value "600".
```
Comment (rickard-von-essen) [inline website/source/docs/builders/oneandone.html.md]: If we are using a string here I think it's better to take a `time.Duration` and convert it to seconds or whatever the API expects.

File: builder/oneandone/config.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	if c.Timeout == 0 {
+		c.Timeout = 600
+	}
+
+	if es := c.Comm.Prepare(&c.ctx); len(es) > 0 {
+		errs = packer.MultiErrorAppend(errs, es...)
+	}
+	c.Comm.SSHPort = 22
```
Comment (rickard-von-essen) [inline builder/oneandone/config.go]: Duplicate port

**Comment by jasminSPC** [comment]
it's being used in this function
```
func (api *API) WaitForState(in ApiInstance, state string, sec time.Duration, count int) error {
```
as parameter `count int`

File: builder/oneandone/ssh.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Error setting up SSH config: %s", err)
+	}
+
+	return &ssh.ClientConfig{
+		User: config.Comm.SSHUsername,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+	}, nil
```
Comment (rickard-von-essen) [inline builder/oneandone/ssh.go]: What about images with passwords?

File: builder/oneandone/config.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+	Comm                communicator.Config `mapstructure:",squash"`
+
+	Token         string `mapstructure:"token"`
+	Url           string `mapstructure:"url"`
+	SSHKey        string
+	SSHKey_path   string              `mapstructure:"ssh_key_path"`
```
Comment (rickard-von-essen) [inline builder/oneandone/config.go]: Is this not the same as `ssh_private_key_file`?

File: builder/oneandone/config.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	SnapshotName  string              `mapstructure:"image_name"`
+	Image         string              `mapstructure:"source_image_name"`
+	ImagePassword string              `mapstructure:"image_password"`
+	DiskSize      int                 `mapstructure:"disk_size"`
+	Timeout       int                 `mapstructure:"timeout"`
```
Comment (rickard-von-essen) [inline builder/oneandone/config.go]: Docs says string

**Comment by rickard-von-essen** [comment]
Nice! I'll try to review this soon.

File: builder/oneandone/config.go (hunk 4 above)
Comment (jasminSPC) [inline builder/oneandone/config.go]: I used it the same way here https://github.com/StackPointCloud/packer/blob/master/builder/profitbricks/config.go#L25 so I thought it was ok to use it again.

File: website/source/docs/builders/oneandone.html.md (hunk 1 above)
Comment (jasminSPC) [inline website/source/docs/builders/oneandone.html.md]: it's being used in this function
```
func (api *API) WaitForState(in ApiInstance, state string, sec time.Duration, count int) error {
```
as parameter `count int`

**Comment by rickard-von-essen** [comment]
All issues in #4169 also applies to this.

**Comment by mwhooker** [comment]
@jasminSPC just to double check, would you be willing to maintain this builder if merged?

**Comment by jasminSPC** [comment]
@mwhooker Yes that is correct

**Comment by rickard-von-essen** [comment]
Awesome!

**Comment by jasminSPC** [comment]
@rickard-von-essen Is there anything else?

**Comment by rickard-von-essen** [comment]
I'm happy but haven't actually tested it. @mwhooker handing this over to you.

**Comment by jasminSPC** [comment]
I'm going to assume you will need credentials for this plugin. So probably we will handle this over the email.

**Pull request merged**